File: hooks/post-oxisym-scan.py
#!/usr/bin/env python3
"""PostToolUse hook (Edit|Write): run oxisym via Dylint to detect structural duplicates in Rust code."""

import json
import os
import subprocess
import sys

from project_root import find_project_root

OXISYM_LIB_DIR = "/home/exp/.claude/external/oxisym/target/release"
DYLINT_BIN_DIR = "/home/exp/.claude/external/dylint/target/release"
NIX_SHELL = "/home/exp/.claude/external/nightly-rust-combined.nix"
FAKE_RUSTUP_DIR = "/tmp"
FAKE_RUSTUP_HOME = "/tmp/fake-rustup-home"
FAKE_CARGO_HOME = "/tmp/fake-cargo-home"
TOOLCHAIN = "nightly-2025-09-18-x86_64-unknown-linux-gnu"
OPENSSL_LIB = "/nix/store/bga5xf95jaypy385hvxm4h3yxl3m1566-openssl-3.6.1/lib"
FENIX_SYSROOT = "/nix/store/gx7i7dg4c0s8g4ycsh0q7bj2w3x9sl2g-rust-mixed"


def ensure_fake_rustup() -> None:
    """Create fake rustup wrapper and sysroot symlink if missing."""
    rustup_path = os.path.join(FAKE_RUSTUP_DIR, "rustup")
    if not os.path.exists(rustup_path):
        with open(rustup_path, "w") as f:
            f.write(f"""#!/bin/sh
case "$1" in
    which)
        case "$2" in
            rustc) echo "{FENIX_SYSROOT}/bin/rustc" ;;
            cargo) echo "{FENIX_SYSROOT}/bin/cargo" ;;
            *) echo "{FENIX_SYSROOT}/bin/rustc" ;;
        esac
        ;;
    show)
        echo "{TOOLCHAIN} (default)"
        ;;
    run)
        shift 2
        exec "$@"
        ;;
    *)
        echo "rustup-compatible wrapper for NixOS" >&2
        ;;
esac
""")
        os.chmod(rustup_path, 0o755)

    toolchain_dir = os.path.join(FAKE_RUSTUP_HOME, "toolchains")
    symlink = os.path.join(toolchain_dir, TOOLCHAIN)
    if not os.path.exists(symlink):
        os.makedirs(toolchain_dir, exist_ok=True)
        os.symlink(FENIX_SYSROOT, symlink)


def main() -> None:
    data = json.load(sys.stdin)
    tool_input = data.get("tool_input", {})
    file_path = tool_input.get("file_path", "") or tool_input.get("path", "")

    if not file_path or not file_path.endswith(".rs"):
        return

    project_root = find_project_root(os.path.dirname(file_path))
    if not project_root:
        return

    ensure_fake_rustup()

    dylint_cmd = (
        f"export PATH={FAKE_RUSTUP_DIR}:{DYLINT_BIN_DIR}:$PATH && "
        f"export LD_LIBRARY_PATH={OPENSSL_LIB}:$LD_LIBRARY_PATH && "
        f"export RUSTUP_TOOLCHAIN={TOOLCHAIN} && "
        f"export RUSTUP_HOME={FAKE_RUSTUP_HOME} && "
        f"export CARGO_HOME={FAKE_CARGO_HOME} && "
        f"export DYLINT_LIBRARY_PATH={OXISYM_LIB_DIR} && "
        f"cargo dylint --lib oxisym 2>&1"
    )

    try:
        result = subprocess.run(
            ["nix-shell", NIX_SHELL, "--run", dylint_cmd],
            capture_output=True,
            text=True,
            cwd=project_root,
            timeout=120,
        )
    except (subprocess.TimeoutExpired, FileNotFoundError):
        return

    output = result.stdout.strip()
    if not output:
        return

    warnings = [
        line for line in output.splitlines()
        if any(k in line for k in ("structural similarity", "same body structure", "same function"))
        or line.strip().startswith("= help:")
    ]

    if not warnings:
        return

    listing = "\n".join(f"  - {w.strip()}" for w in warnings)
    json.dump(
        {
            "hookSpecificOutput": {
                "hookEventName": "PostToolUse",
                "additionalContext": (
                    f"oxisym が構造的複製を検出しました。\n{listing}\n"
                    "共通化または抽象化を検討してください。"
                ),
            }
        },
        sys.stdout,
    )


if __name__ == "__main__":
    main()

File: hooks-rs/src/hooks/post_oxisym_scan.rs
use std::path::Path;
use std::process::Command;

use crate::input::HookInput;
use crate::output::stop;
use crate::project_root::find_project_root;

const OXISYM_LIB_DIR: &str = "/home/exp/.claude/external/oxisym/target/release";
const DYLINT_BIN_DIR: &str = "/home/exp/.claude/external/dylint/target/release";
const NIX_SHELL: &str = "/home/exp/.claude/external/nightly-rust-combined.nix";
const TOOLCHAIN: &str = "nightly-2025-09-18-x86_64-unknown-linux-gnu";
const OPENSSL_LIB: &str = "/nix/store/bga5xf95jaypy385hvxm4h3yxl3m1566-openssl-3.6.1/lib";
const FENIX_SYSROOT: &str = "/nix/store/gx7i7dg4c0s8g4ycsh0q7bj2w3x9sl2g-rust-mixed";
const FAKE_RUSTUP_DIR: &str = "/tmp";
const FAKE_RUSTUP_HOME: &str = "/tmp/fake-rustup-home";
const FAKE_CARGO_HOME: &str = "/tmp/fake-cargo-home";

pub fn run(input: &HookInput) {
    let file_path = input.tool_input.file_path_resolved();
    if !file_path.ends_with(".rs") {
        return;
    }

    let path = Path::new(file_path);
    if !path.is_file() {
        return;
    }

    let project_root = match find_project_root(path, "Cargo.toml") {
        Some(root) => root,
        None => return,
    };

    ensure_fake_rustup();

    let output = match run_oxisym(&project_root) {
        Ok(out) => out,
        Err(_) => return,
    };

    let warnings = parse_warnings(&output);
    if warnings.is_empty() {
        return;
    }

    let listing = warnings
        .iter()
        .map(|w| format!("  - {w}"))
        .collect::<Vec<_>>()
        .join("\n");
    stop(&format!(
        "oxisym が構造的複製を検出しました。\n{listing}\n共通化または抽象化を検討してください。"
    ));
}

fn ensure_fake_rustup() {
    let rustup_path = format!("{FAKE_RUSTUP_DIR}/rustup");
    if !Path::new(&rustup_path).exists() {
        let script = format!(
            "#!/bin/sh\ncase \"$1\" in\n  which) case \"$2\" in \
             rustc) echo \"{FENIX_SYSROOT}/bin/rustc\" ;; \
             cargo) echo \"{FENIX_SYSROOT}/bin/cargo\" ;; \
             *) echo \"{FENIX_SYSROOT}/bin/rustc\" ;; esac ;;\n  \
             show) echo \"{TOOLCHAIN} (default)\" ;;\n  \
             run) shift 2; exec \"$@\" ;;\n  \
             *) echo \"rustup-compatible wrapper for NixOS\" >&2 ;;\nesac\n"
        );
        let _ = std::fs::write(&rustup_path, script);
        let _ = std::process::Command::new("chmod")
            .args(["+x", &rustup_path])
            .status();
    }

    let toolchain_dir = format!("{FAKE_RUSTUP_HOME}/toolchains");
    let symlink = format!("{toolchain_dir}/{TOOLCHAIN}");
    if !Path::new(&symlink).exists() {
        let _ = std::fs::create_dir_all(&toolchain_dir);
        let _ = std::os::unix::fs::symlink(FENIX_SYSROOT, &symlink);
    }
}

fn run_oxisym(project_root: &Path) -> Result<String, String> {
    let inner_cmd = format!(
        "export PATH={FAKE_RUSTUP_DIR}:{DYLINT_BIN_DIR}:$PATH && \
         export LD_LIBRARY_PATH={OPENSSL_LIB}:$LD_LIBRARY_PATH && \
         export RUSTUP_TOOLCHAIN={TOOLCHAIN} && \
         export RUSTUP_HOME={FAKE_RUSTUP_HOME} && \
         export CARGO_HOME={FAKE_CARGO_HOME} && \
         export DYLINT_LIBRARY_PATH={OXISYM_LIB_DIR} && \
         cargo dylint --lib oxisym 2>&1"
    );

    let output = Command::new("nix-shell")
        .args([NIX_SHELL, "--run", &inner_cmd])
        .current_dir(project_root)
        .output()
        .map_err(|e| e.to_string())?;

    let stdout = String::from_utf8_lossy(&output.stdout).to_string();
    let stderr = String::from_utf8_lossy(&output.stderr).to_string();
    Ok(format!("{stdout}\n{stderr}"))
}

fn parse_warnings(output: &str) -> Vec<String> {
    let mut warnings = Vec::new();
    for line in output.lines() {
        let trimmed = line.trim();
        if trimmed.contains("structural similarity")
            || trimmed.contains("same body structure")
            || trimmed.contains("same function")
        {
            warnings.push(trimmed.to_string());
        }
        if trimmed.starts_with("= help:") {
            warnings.push(trimmed.to_string());
        }
    }
    warnings
}
